File: src/main/java/twao/assigners/NobleAssigner.java
package twao.assigners;

import twao.villages.AllyVillage;
import twao.villages.TargetVillage;
import twao.villages.Village;
import twao.VillageAssignment;

import java.util.*;

public class NobleAssigner implements Runnable {
    private List<TargetVillage> targets;
    private List<AllyVillage> attackingVillages;
    private Village relativityPoint;
    private final int maxNobleRange;
    private final boolean assigningFakes;

    public NobleAssigner(List<TargetVillage> targets, List<AllyVillage> attackingVillages, Village relativityPoint, boolean assigningFakes, int maxNobleRange) {
        this.targets = targets;
        this.attackingVillages = attackingVillages;
        this.relativityPoint = relativityPoint;
        this.maxNobleRange = (int) Math.pow(maxNobleRange, 2) -1;
        this.assigningFakes = assigningFakes;
    }

    private void sortTargets() {
        for (Village vil : targets) {
            vil.setRelativeDistance(relativityPoint);
        }

        targets.sort(Comparator.comparing(Village::getRelativeDistance));
    }

    @Override
    public void run() {
        List<AllyVillage> processingList = new LinkedList<>();

        List<AllyVillage> uselessVillages;
        List<VillageAssignment> assignmentsList;
        AllyVillage closestVil;
        int distance;

        processingList.addAll(attackingVillages);

        sortTargets();
        for (TargetVillage target : targets) {
            while (target.isAssignCompleted() == false) {
                if (processingList.size() == 0) {
                    break;
                }

                //init closestVil and distance
                closestVil = processingList.get(0);
                distance = target.getDistanceTo(closestVil);

                //search for closest village
                uselessVillages = new LinkedList<>();
                for (AllyVillage attacker : processingList) {
                    if (attacker.getOwner().hasNoble() == false) {
                        uselessVillages.add(attacker);
                    }

                    else if (attacker.getDistanceTo(target) < distance) {
                        closestVil = attacker;
                        distance = closestVil.getDistanceTo(target);
                    }
                }
                processingList.removeAll(uselessVillages);

                if (maxNobleRange < closestVil.getRelativeDistance()) {
                    break;
                }

                if ( closestVil.getOwner().hasNoble() ) { //make sure that closestVil can send noble
                    if (assigningFakes) {
                        assignmentsList = closestVil.getOwner().getFakeNobleAssignments();
                    }
                    else {
                        assignmentsList = closestVil.getOwner().getNobleAssignments();
                    }

                    assignmentsList.add( new VillageAssignment(closestVil, target, distance) );

                    closestVil.getOwner().delegateNoble();

                    attackingVillages.remove(closestVil);
                    processingList.remove(closestVil);

                    target.attack();
                } 
                else {
                    processingList.remove(closestVil);
                }
            }
        }
    }
}

File: src/main/java/twao/assigners/NobleAssigner.kt
package twao.assigners

import twao.Player
import twao.villages.AllyVillage
import twao.villages.TargetVillage
import twao.villages.Village
import java.util.*
import kotlin.Comparator

class NobleAssigner internal constructor(
        targets: MutableList<TargetVillage>,
        resources: MutableList<AllyVillage>,
        mainReferencePoint: Village,
        isAssigningFakes: Boolean,
        private val maxNobleRange: Int
) : Assigner (targets, resources, mainReferencePoint, isAssigningFakes) {

    override val targetsQueue = PriorityQueue< Pair<TargetVillage, Int> >(targets.size, setTargetsPriorities())
    override val resourcesQueue = PriorityQueue< Pair<AllyVillage, Int> >(resources.size, setResourcesPriorities())

    override val offAction = Player::putNobleAssignment
    override val fakeAction = Player::putFakeNobleAssignment

    private var resWithNobles = resources.filter { it.owner.hasNoble() }
                                         .toMutableList()

    override fun putResourcesToQueue(referencePoint: Village) {
        resWithNobles.forEach { resourcesQueue.offer( Pair(it, Village.distance(it, referencePoint)) ) }
    }

    override fun run() {
        putTargetsToQueue(referencePoint = mainReferencePoint)

        for ((target, _) in targetsQueue) {
            if (resWithNobles.isEmpty()) break

            handleTarget(target)
        }

    }

    private fun handleTarget(target: TargetVillage) {
        putResourcesToQueue(referencePoint = target)

        for ((nearestAllyVillage, distance) in resourcesQueue) {
            if (target.isAssignCompleted())
                break
            if (distance > maxNobleRange)
                break

            assign(nearestAllyVillage, target, distance)
            resources.remove(nearestAllyVillage)

            nearestAllyVillage.owner.delegateNoble()
            if (!nearestAllyVillage.owner.hasNoble()) updateResWithNobles()
            else resWithNobles.remove(nearestAllyVillage)

            target.attack()
        }

        if (target.isAssignCompleted()) {
            targets.remove(target)
        }
    }

    private fun updateResWithNobles() {
        resWithNobles = resWithNobles.filter { it.owner.hasNoble() }
                                     .toMutableList()
    }

    private fun setTargetsPriorities()
            = Comparator.comparing(Pair<Village, Int>::second)

    private fun setResourcesPriorities()
            = Comparator.comparing(Pair<Village, Int>::second)
}